### PR TITLE
Team reads migration

### LINKED ISSUE #4630: Read Team from PG instead of RethinkDB (Phase 2 of 3)
After #4629 is working is prod, we need to start reading from pg instead of RethinkDB. See #4530 

AC
- [ ] All reads for the `Team` table come from pg, not RethinkDB
- [ ] if you nuke the data in the Team table in rethinkdb, nothing bad happens

Estimate: 16 hours

### PR DESCRIPTION
**Pull request by tianrunhe**
Resolves #4630.

I brought the changes from Tiffany's last PR #4817 but also resolves the merge conflicts and update a lot of the code.

Tests:
- [x] Run through the [release test checklist](https://www.notion.so/parabol/Release-Tests-Checklist-a4146ee7b24f455681e6ec58225974df)
- [x] Manually test the following mutations in `/admin/graphql`
  - [x] `archiveOrganization`
  - [x] `downgradeToPersonal`
  - [x] `updateCreditCard`
  - [x] `updateTeamName`
  - [x] `upgradeToPro`
  - [x] `hideConversionModal`
  - [x]  `moveTeamToOrg`
  - [x] `startRetrospective`
  - [x] `verifiedInvitation`
  - [x] `acceptTeamInvitation`

### CODE REVIEW

File: packages/server/postgres/queries/getTeamByTeamId.ts
Comment (tiffanyhan) [inline packages/server/postgres/queries/getTeamByTeamId.ts:7]: curious why add the bang operator?  what happens if `catchAndLog` returns null? should we add `teams === null ? [] : teams` before returning the first value, similarly to `packages/server/postgres/queries/getTeamsByOrgId.ts`?

**Review by tiffanyhan — COMMENTED** [review]
wow, i'm glad you were able to use work from that old PR! i hope it wasn't too hairy to update 😬 thanks for making these changes!

File: packages/server/postgres/queries/getTeamByTeamId.ts
Comment (tianrunhe) [reply packages/server/postgres/queries/getTeamByTeamId.ts:7]: This method is expected to return a single team object, isn't it? I did have difficulty in handling the nullify here. Any better idea?

Comment (tiffanyhan) [reply packages/server/postgres/queries/getTeamByTeamId.ts:7]: i think we could do 

```
teams === null ? [] : teams
return teams[0]
```

if the query returns null, then this wrapper will return undefined instead of raising an error

File: packages/server/utils/authorization.ts
Comment (mattkrick) [inline packages/server/utils/authorization.ts:104]: let's nuke this thing!

File: packages/server/postgres/queries/getTeamsByOrgId.ts
Comment (mattkrick) [inline packages/server/postgres/queries/getTeamsByOrgId.ts:18]: let's coerce `isArchived` to a boolean here. then we can get rid of the CAST + condition is the SQL.

File: packages/server/postgres/queries/getTeamByTeamId.ts
Comment (mattkrick) [reply packages/server/postgres/queries/getTeamByTeamId.ts:7]: let's nuke this file & just use `getTeamsById`

File: packages/server/graphql/mutations/moveTeamToOrg.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/moveTeamToOrg.ts:21]: remove `r({org: ` wrapper

File: packages/server/graphql/mutations/helpers/resolveDowngradeToPersonal.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/resolveDowngradeToPersonal.ts:22]: let's not return `teamIds` here so we can remove the RETURN from our write query.
then, in https://github.com/ParabolInc/parabol/blob/36a79f894e9a78622a4fa95c8aa308591c8f1e6c/packages/server/graphql/mutations/downgradeToPersonal.ts#L44 we can just get the teamIds via the new dataLoader

File: packages/server/graphql/mutations/addTeam.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/addTeam.ts:44]: so much cleaner! :tada:

**Review by mattkrick — CHANGES_REQUESTED** [review]
great work! mostly just little things

File: packages/server/graphql/mutations/startRetrospective.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/startRetrospective.ts:64]: be defensive! if they pass in "foo" as the teamId this will be undefined & throw.

File: packages/server/graphql/mutations/downgradeToPersonal.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/downgradeToPersonal.ts:45]: this code was not tested! this dataloader will not return an object containing `teamIds`.
can you please:
- add a `GQLContext` type to the context object at the top so we get some type safety
- add a checklist of tests that you performed (see https://github.com/ParabolInc/parabol/pull/5205 as an example)

**Comment by mattkrick** [comment]
tagging @tiffanyhan to perform the code review

**Comment by tiffanyhan** [comment]
@bruce i ran into this unexpected behavior while going through the tests:

- [ ] start with a team on a free org
- [ ] start retrospective meeting, but don't end it or go to discuss yet
- [ ] set `showConversionModal=true` on the meeting object in rethinkdb
- [ ] now upgrade the org to pro
- [ ] see in rethink db that `showConversionModal=true` even though it should be false by `hideConversionModal`

Can you please run through the reproduction steps and see if it's behaving similarly for you? If so, I can just open up a new issue that won't block this PR, since I see this PR doesn't have any changes with regard to the file `packages/server/graphql/mutations/helpers/hideConversionModal.ts`.

File: packages/server/graphql/intranetSchema/mutations/backupOrganization.ts
Comment (tiffanyhan) [inline packages/server/graphql/intranetSchema/mutations/backupOrganization.ts:56]: can we make this plural? i.e. team => teams

**Review by tiffanyhan — CHANGES_REQUESTED** [review]

**Comment by tianrunhe** [comment]
> @tianrunhe i ran into this unexpected behavior while going through the tests:
> 
> *   [ ] start with a team on a free org
> *   [ ] start retrospective meeting, but don't end it or go to discuss yet
> *   [ ] set `showConversionModal=true` on the meeting object in rethinkdb
> *   [ ] now upgrade the org to pro
> *   [ ] see in rethink db that `showConversionModal=true` even though it should be false by `hideConversionModal`
> 
> Can you please run through the reproduction steps and see if it's behaving similarly for you? If so, I can just open up a new issue that won't block this PR, since I see this PR doesn't have any changes with regard to the file `packages/server/graphql/mutations/helpers/hideConversionModal.ts`.

@tiffanyhan I can reproduce this and I think it's due to these two lines:

[https://github.com/ParabolInc/parabol/blob/master/packages/server/graphql/mutations/helpers/hideConversionModal.ts#L7-L8](https://github.com/ParabolInc/parabol/blob/master/packages/server/graphql/mutations/helpers/hideConversionModal.ts#L7-L8)

Which makes me question if this is a valid test case, i.e., this could never happen in real world:
1. `showConversionModal` can be attributes on both `Organization` and `Meeting` object
2. For `Organization` object, its value can only be mutated through the **private** mutation `flagConversionModal`
3. The `showConversionModal` attribute for a `Meeting` object is set when the meeting starts, and it checks the [`showConversionModal` for the organization the team is in](https://github.com/ParabolInc/parabol/blob/master/packages/server/graphql/mutations/startRetrospective.ts#L69)

Therefore there won't be a case where the attribute for the org is false whereas a meeting in that org has the attribute value to be true, as specified in the test case.

**Review by tiffanyhan — APPROVED** [review]
lgtm! ran through all specified mutations and all worked well

File: packages/server/postgres/queries/getTeamsByOrgIds.ts
Comment (mattkrick) [inline packages/server/postgres/queries/getTeamsByOrgIds.ts:10]: i think you want to omit `orgIds` here?
might just be simpler to write: options: {isArchived?: boolean} = {}

File: packages/server/postgres/queries/updateTeamByOrgId.ts
Comment (mattkrick) [inline packages/server/postgres/queries/updateTeamByOrgId.ts:12]: no `return await` (i'll turn on a rule after we get this merged)
remove catchAndLog.

File: packages/server/safeMutations/acceptTeamInvitation.ts
Comment (mattkrick) [inline packages/server/safeMutations/acceptTeamInvitation.ts:64]: here you're saying team could be null, but in the next line you're accessing team. that'll cause a nasty runtime error if it really is null. (there are a handful of other places this happens)

File: packages/server/safeMutations/safeArchiveTeam.ts
Comment (mattkrick) [inline packages/server/safeMutations/safeArchiveTeam.ts:25]: i'm not sure this was checked, `pgResult` is going to be an empty array...

File: packages/server/safeMutations/safeArchiveTeam.ts
Comment (mattkrick) [inline packages/server/safeMutations/safeArchiveTeam.ts:50]: `updateTeamByTeamId` still has a catchAndLog in it (as do a few other files related to the Team table)
could you clean up those remaining files?

File: packages/server/graphql/mutations/addTeam.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/addTeam.ts:43]: So...I can't create a team on this PR :astonished: 

calling `addTeam` throws an error because getTeamsByOrgIds takes an array as the first param. 
While the fix is easy, I'm a little concerned that this broken code went through 2 reviews approvals.
In the future can we be sure to test touched files? Some things are sure to slip through the cracks, and that's OK! But if 3 people test all the changed files we should be able to avoid broken stuff landing in prod.

@tianrunhe @tiffanyhan

Comment (tianrunhe) [reply packages/server/graphql/mutations/addTeam.ts:43]: This is my bad. Apologize! Here's the retro:  
  
I checked the references for `getTeamsByOrgId` and saw two patterns  
1. places where `getTeamsByOrgIds` gets called directly
2. it's used in `teamsByOrgId` as a foreign data loader and the data loader is then used in a lot of places.

I tested everything in 2. and that's how I build the test cases. I did not go through all use cases in 1. and that's how `addTeam` get slipped.

In the future I think we do not only need to test all touched files, but also references to changed methods.

**Comment by mattkrick** [comment]
Next steps for dataloaders:
- remove `teams` primary loader
- remove all foreignLoaders that depend on the `teams` primary loader ('teamsByOrgId')
- add a `teamsByOrgId` customLoader

**Comment by tianrunhe** [comment]
@mattkrick I've updated the loaders, please review. Thanks!

File: packages/server/safeMutations/safeArchiveTeam.ts
Comment (mattkrick) [inline packages/server/safeMutations/safeArchiveTeam.ts:52]: The `team` object that comes from RethinkDB is null if the object wasn't changed.
`RETURNING *` will always return a team object as long as it exists, so the error message in `archiveTeam` would be wrong.
This is also the only place that `RETURNING *` is used, so it feels a little bad that we're always performing an extra read just for this 1 place.

Options:
- write an `archiveTeam` query that includes `WHERE "isArchived" = false`
- query the team at the top of this function. if `isArchived == false` then return early with {team: null}. else, query the team again (maybe via dataLoader). There's a small chance of a race condition with this, but it's probably OK.
-  ???

**Comment by mattkrick** [comment]
Just one query to go!

**Comment by mattkrick** [comment]
LGTM!

**Pull request merged**
